File: src/apps/bench/cppnetbenchs/benchs/budp.cc
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2021 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>
#include <base/Panic.h>
#include <base/time/Profile.h>

#include <m3/Test.h>
#include <m3/com/Semaphore.h>
#include <m3/net/UdpSocket.h>
#include <m3/session/NetworkManager.h>
#include <m3/stream/Standard.h>
#include <m3/vfs/Waiter.h>

#include "../cppnetbenchs.h"

using namespace m3;

static Option<size_t> send_recv(FileWaiter &waiter, FileRef<UdpSocket> &socket,
                                const Endpoint &dest, const uint8_t *send_buf, size_t sbuf_size,
                                TimeDuration timeout, uint8_t *recv_buf, size_t rbuf_size) {
    socket->send_to(send_buf, sbuf_size, dest);

    waiter.wait_for(timeout);

    if(socket->has_data())
        return socket->recv(recv_buf, rbuf_size);
    return None;
}

NOINLINE static void latency() {
    const TimeDuration TIMEOUT = TimeDuration::from_secs(1);

    NetworkManager net("net");

    uint8_t request[1024];
    uint8_t response[1024];

    auto socket = UdpSocket::create(net);

    const size_t samples = 15;
    Endpoint src;
    Endpoint dest = Endpoint(IpAddr(192, 168, 112, 1), 1337);

    FileWaiter waiter;
    waiter.add(socket->fd(), File::INPUT);

    // do one initial send-receive with a higher timeout than the smoltcp-internal timeout to
    // workaround the high ARP-request delay with the loopback device.
    send_recv(waiter, socket, dest, request, 1, TimeDuration::from_secs(6), response,
              sizeof(response));

    size_t warmup = 5;
    while(warmup--)
        send_recv(waiter, socket, dest, request, 8, TIMEOUT, response, sizeof(response));

    const size_t packet_size[] = {8, 16, 32, 64, 128, 256, 512, 1024};

    for(auto pkt_size : packet_size) {
        Results<TimeDuration> res(samples);

        while(res.runs() < samples) {
            auto start = TimeInstant::now();

            size_t recv_len = send_recv(waiter, socket, dest, request, pkt_size, TIMEOUT, response,
                                        sizeof(response))
                                  .unwrap();

            auto stop = TimeInstant::now();

            WVASSERTEQ(recv_len, pkt_size);

            auto duration = stop.duration_since(start);
            println("RTT ({}b): {} us"_cf, pkt_size, duration.as_micros());
            res.push(duration);
        }

        auto name = OStringStream();
        format_to(name, "network latency ({}b)"_cf, pkt_size);
        WVPERF(name.str(), MilliFloatResultRef<TimeDuration>(res));
    }
}

NOINLINE static void bandwidth() {
    NetworkManager net("net");

    auto socket = UdpSocket::create(
        net, DgramSocketArgs().send_buffer(8, 64 * 1024).recv_buffer(32, 256 * 1024));

    constexpr size_t packet_size = 1024;

    uint8_t request[1024];
    uint8_t response[1024];

    Endpoint dest = Endpoint(IpAddr(192, 168, 112, 1), 1337);

    size_t warmup = 5;
    size_t packets_to_send = 105;
    size_t packets_to_receive = 100;
    size_t burst_size = 2;
    TimeDuration timeout = TimeDuration::from_secs(1);

    size_t packet_sent_count = 0;
    size_t packet_received_count = 0;
    size_t received_bytes = 0;

    FileWaiter waiter;
    waiter.add(socket->fd(), File::INPUT | File::OUTPUT);

    while(warmup--)
        send_recv(waiter, socket, dest, request, 8, timeout, response, sizeof(response));

    socket->set_blocking(false);

    auto start = TimeInstant::now();
    auto last_received = start;
    size_t failures = 0;
    while(true) {
        // Wait for wakeup (message or credits received)
        if(failures >= 10) {
            failures = 0;
            auto waited = TimeInstant::now().duration_since(last_received);
            if(waited > timeout)
                break;
            if(packet_sent_count >= packets_to_send) {
                // we are not interested in output anymore
                waiter.remove(socket->fd());
                waiter.add(socket->fd(), File::INPUT);
            }
            waiter.wait_for(timeout - waited);
        }

        size_t send_count = burst_size;
        while(send_count-- && packet_sent_count < packets_to_send) {
            if(socket->send_to(request, packet_size, dest)) {
                packet_sent_count++;
                failures = 0;
            }
            else {
                failures++;
                break;
            }
        }

        size_t receive_count = burst_size;
        while(receive_count--) {
            if(auto pkt_size = socket->recv(response, sizeof(response))) {
                received_bytes += pkt_size.unwrap();
                packet_received_count++;
                last_received = TimeInstant::now();
                failures = 0;
            }
            else {
                failures++;
                break;
            }
        }

        if(packet_received_count >= packets_to_receive)
            break;
    }

    println("Benchmark done."_cf);

    println("Sent packets: {}"_cf, packet_sent_count);
    println("Received packets: {}"_cf, packet_received_count);
    println("Received bytes: {}"_cf, received_bytes);
    auto duration = last_received.duration_since(start);
    println("Duration: {}"_cf, duration);
    auto secs = static_cast<float>(duration.as_nanos()) / 1000000000.f;
    float mbps = (static_cast<float>(received_bytes) / secs) / (1024 * 1024);

    auto res = OStringStream();
    format_to(res, "{}  MiB/s (+/- 0 with 1 runs)\n"_cf, mbps);
    WVPERF("network bandwidth", res.str());
}

void budp() {
    // wait for UDP socket just once
    Semaphore::attach("net-udp").down();

    RUN_BENCH(latency);
    RUN_BENCH(bandwidth);
}

File: src/apps/bench/rustnetbenchs/src/budp.rs
/*
 * Copyright (C) 2021 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use m3::com::Semaphore;
use m3::errors::{Code, Error};
use m3::format;
use m3::net::{DGramSocket, DgramSocketArgs, Endpoint, Socket, UdpSocket};
use m3::println;
use m3::session::NetworkManager;
use m3::test::WvTester;
use m3::time::{Results, TimeDuration, TimeInstant};
use m3::vfs::{File, FileEvent, FileRef, FileWaiter};
use m3::{wv_assert_eq, wv_assert_ok, wv_perf, wv_run_test};

const TIMEOUT: TimeDuration = TimeDuration::from_secs(1);

pub fn run(t: &mut dyn WvTester) {
    // wait once for UDP, because it's connection-less
    wv_assert_ok!(Semaphore::attach("net-udp").unwrap().down());

    wv_run_test!(t, latency);
    wv_run_test!(t, bandwidth);
}

fn send_recv(
    waiter: &mut FileWaiter,
    socket: &mut FileRef<UdpSocket>,
    dest: Endpoint,
    msg: &mut [u8],
    timeout: TimeDuration,
) -> Result<usize, Error> {
    wv_assert_ok!(socket.send_to(msg, dest));

    waiter.wait_for(timeout);

    if socket.has_data() {
        socket.recv(msg)
    }
    else {
        Err(Error::new(Code::Timeout))
    }
}

fn latency(t: &mut dyn WvTester) {
    let nm = wv_assert_ok!(NetworkManager::new("net"));
    let mut socket = wv_assert_ok!(UdpSocket::new(DgramSocketArgs::new(nm)));

    wv_assert_ok!(socket.set_blocking(false));

    let samples = 5;
    let dest = Endpoint::new(crate::DST_IP.get(), 1337);

    let mut buf = [0u8; 1024];

    let mut waiter = FileWaiter::default();
    waiter.add(socket.fd(), FileEvent::INPUT);

    // do one initial send-receive with a higher timeout than the smoltcp-internal timeout to
    // workaround the high ARP-request delay with the loopback device.
    wv_assert_ok!(send_recv(
        &mut waiter,
        &mut socket,
        dest,
        &mut buf,
        TimeDuration::from_secs(6)
    ));

    // warmup
    for _ in 0..5 {
        // ignore failures here
        send_recv(&mut waiter, &mut socket, dest, &mut buf, TIMEOUT).ok();
    }

    let packet_sizes = [8, 16, 32, 64, 128, 256, 512, 1024];

    for pkt_size in &packet_sizes {
        let mut res = Results::new(samples);

        for _ in 0..samples {
            let start = TimeInstant::now();

            if let Ok(recv_size) = send_recv(
                &mut waiter,
                &mut socket,
                dest,
                &mut buf[0..*pkt_size],
                TIMEOUT,
            ) {
                let stop = TimeInstant::now();

                wv_assert_eq!(t, *pkt_size, recv_size);
                res.push(stop.duration_since(start));
            }
        }

        wv_perf!(
            format!("network latency ({}b)", pkt_size),
            format!(
                "{:.4} ms (+/- {:.4} ms with {} runs)",
                res.avg().as_nanos() as f32 / 1e6,
                res.stddev().as_nanos() as f32 / 1e6,
                res.runs()
            )
        );
    }
}

fn bandwidth(t: &mut dyn WvTester) {
    const PACKETS_TO_SEND: usize = 105;
    const PACKETS_TO_RECEIVE: usize = 100;
    const BURST_SIZE: usize = 2;
    const TIMEOUT: TimeDuration = TimeDuration::from_secs(1);

    let nm = wv_assert_ok!(NetworkManager::new("net"));
    let mut socket = wv_assert_ok!(UdpSocket::new(
        DgramSocketArgs::new(nm)
            .send_buffer(8, 64 * 1024)
            .recv_buffer(32, 256 * 1024)
    ));
    wv_assert_ok!(socket.set_blocking(false));

    let dest = Endpoint::new(crate::DST_IP.get(), 1337);

    let mut buf = [0u8; 1024];

    let mut waiter = FileWaiter::default();
    waiter.add(socket.fd(), FileEvent::INPUT | FileEvent::OUTPUT);

    for _ in 0..10 {
        // ignore errors here
        send_recv(&mut waiter, &mut socket, dest, &mut buf, TIMEOUT).ok();
    }

    let start = TimeInstant::now();
    let mut timeout = start + TIMEOUT;
    let mut sent_count = 0;
    let mut receive_count = 0;
    let mut received_bytes = 0;
    let mut failures = 0;

    loop {
        if failures > 9 {
            failures = 0;
            let rem = timeout.checked_duration_since(TimeInstant::now());
            match rem {
                Some(d) => {
                    if sent_count >= PACKETS_TO_SEND {
                        // we are not interested in output anymore
                        waiter.remove(socket.fd());
                        waiter.add(socket.fd(), FileEvent::INPUT);
                    }
                    waiter.wait_for(d);
                },
                None => break,
            }
        }

        for _ in 0..BURST_SIZE {
            if sent_count >= PACKETS_TO_SEND {
                break;
            }

            match socket.send_to(&buf, dest) {
                Err(e) => {
                    wv_assert_eq!(t, e.code(), Code::WouldBlock);
                    failures += 1;
                },
                Ok(_) => {
                    sent_count += 1;
                    failures = 0;
                },
            }
        }

        for _ in 0..BURST_SIZE {
            match socket.recv(&mut buf) {
                Err(e) => {
                    wv_assert_eq!(t, e.code(), Code::WouldBlock);
                    failures += 1;
                },
                Ok(size) => {
                    received_bytes += size;
                    receive_count += 1;
                    timeout = TimeInstant::now() + TIMEOUT;
                    failures = 0;
                },
            }
        }

        if receive_count >= PACKETS_TO_RECEIVE {
            break;
        }
    }

    println!("Sent packets: {}", sent_count);
    println!("Received packets: {}", receive_count);
    println!("Received bytes: {}", received_bytes);
    let last_received = timeout - TIMEOUT;
    let duration = last_received.duration_since(start);
    println!("Duration: {:?}", duration);
    let secs = (duration.as_nanos() as f64) / 1e9;
    let mbps = (received_bytes as f64 / secs) / (1024f64 * 1024f64);
    wv_perf!(
        "UDP bandwidth",
        format!("{} MiB/s (+/- 0 with 1 runs)", mbps)
    );
}
